Improved log uploading logic

jlurgo opened the pull request:
Cherry picked changes from 
https://github.com/transitiverobotics/transitive/pull/92

jlurgo commented:
OK, now I'm seeing what I was expecting:
![image](https://github.com/user-attachments/assets/d6da3de9-0bcb-4176-a14c-da6eddfaa5bd)

chfritz commented on robot-agent/logMonitor.js:
```diff
@@ -182,89 +227,94 @@ class LogMonitor {
+      log.debug('No pending logs to process');
+      return;
+    }
+    this.restartUploadingLogs(nextTimeOut); // Restart the log uploading process
```
Why is it better to call this function back using a `setTimeout(..., 0)` rather than just use a while loop to empty the `pendingLogs`?

chfritz reviewed (commented):
Looks good. Just a couple of comments.

jlurgo commented:
I addressed your review @chfritz 

Only thing left (not sure if it's a problem)
I developed it so on mqtt errors it'd retry uploading after 5 secs
To simulate being offline I killed the mosquitto container, but that's not making mqttClient.publish to fail, it waits there forever until I re create the container, when it succeeds and then all pending messages are sent.
Is that ok??

chfritz commented:
> To simulate being offline I killed the mosquitto container, but that's not making mqttClient.publish to fail, it waits there forever until I re create the container, when it succeeds and then all pending messages are sent.
Is that ok??

Isn't that good? I believe it actually might start failing once the mqtt client realizes it is indeed offline. I think the timeout is 1m 10s or similar. Did you wait that long? If not, then I think it would be good to test that case also.

jlurgo commented:
Well I added a short delay between each message and now it works:
![image](https://github.com/user-attachments/assets/9fbf5f95-7f4c-41b8-bebe-b0fc968ce4d8)
I think I get it what the problem is. We get the confirmation from the broker, not from the portal. I think the portal is taking some more time to reconnect to the broker (don't know why the agent would reconnect earlier) and meanwhile, if we send all messages in a burst without delay, all first messages are lost...
Well if this is true I'd expect a fixed time gap of messages after losing connectivity, but I don't see it like that.
Anyways, maybe we need to detect disconnections and resume sending logs only after a heartbeat or something like that.
Adding a delay between messages mitigates the problem a lot though

jlurgo commented:
I did some changes, to detect mqtt going offline, pausing logging, and then resume on the next heartbeat, I think is the right direction, but I'll need to hook to a portal heartbeat instead of a broker one, then it'll be done.
To test I tried to broadcast a HB from the portal to all listeners, but I failed to receive it on the agent... I'll continue trying tomorrow unless you have a better advise @chfritz

chfritz commented:
Let's not use heartbeats from the portal. This is getting too complicated. Like I said, I'm not too worried about the cloud being intermittently offline because that barely ever happens. Your test for the robot being offline by stopping the broker is not a very accurate simulation, because the portal would not lose connection to the broker if one of the robots goes offline. Right now, with the heartbeats, you are trying to fix a problem that only exists in your test setup. The better test might be to run the robot agent in a docker container and simulating that container going offline using `tc` (linux command), e.g., `tc qdisc replace dev eth0 root netem loss random 100` (simulate 100% network loss), or similar.

jlurgo commented:
OK, makes sense, I went back one commit!

chfritz commented on robot-agent/logMonitor.js:
```diff
@@ -115,58 +146,72 @@ class LogMonitor {
+      log.warn('Package is not being watched:', packageName,
+        'Log line will not be parsed:', line);
+      return null; // Skip if package is not being watched
+    }
```
I don't understand this section. If the package is not being watched then why was this function even called? Also, why use `_.get` instead of `this.watchedPackages[packageName]` like everywhere else? Was this perhaps suggested by copilot?

jlurgo commented:
@chfritz I implemented what we discussed about batch uploading
there's the MAX_LOGS_PER_BATCH constant which will need to be tuned, not based on size but only on count, I can improve it

Some thoughts:
- currently we will only batch in case of an offline period, on startup or if logs generation is faster than upload.
- would be good to wait some time between log uploads so to give it some time for logs to accumulate?
